From 6c3953f41ac6bc429fd7b49677998d76b1b4135f Mon Sep 17 00:00:00 2001
From: Kris Shannon <k.shannon@amaze.com.au>
Date: Fri, 26 May 2017 08:43:32 +1000
Subject: [PATCH] MANTA-2409 Muskie pagination subject to Moray filter
 injection MANTA-2803 Cannot mrmdir or mls in manta folder with parentheses in
 name

---
 lib/common.js | 86 +++++++++++++++++++++++++++------------------------
 package.json  |  1 +
 2 files changed, 47 insertions(+), 40 deletions(-)

diff --git a/lib/common.js b/lib/common.js
index b2d72ec..a228168 100644
--- a/lib/common.js
+++ b/lib/common.js
@@ -17,6 +17,7 @@ var httpSignature = require('http-signature');
 
 var assert = require('assert-plus');
 var libmanta = require('libmanta');
+var morayFilter = require('moray-filter');
 var vasync = require('vasync');
 var restify = require('restify');
 var VError = require('verror');
@@ -599,7 +600,16 @@ function readdir(dir, req) {
     // We want the really low-level API here, as we want to go hit the place
     // where all the keys are, not where the dirent itself is.
     var client = req.moray;
-    var filter = '(&';
+    var filter = new morayFilter.AndFilter();
+
+    filter.addFilter(new morayFilter.EqualityFilter({
+        attribute: 'owner',
+        value: account
+    }));
+    filter.addFilter(new morayFilter.EqualityFilter({
+        attribute: 'dirname',
+        value: dir
+    }));
 
     //The 'dir' above comes in as the path of the request.  The 'dir'
     // and 'obj' parameters are filters.
@@ -608,52 +618,19 @@ function readdir(dir, req) {
     var hasObj = (req.params.obj !== undefined ||
                   req.params.object !== undefined);
 
-    if ((hasDir && hasObj) || !(hasDir || hasObj)) {
-        filter += sprintf('(owner=%s)(dirname=%s)', account, dir);
-    } else if (hasDir) {
-        filter += sprintf('(owner=%s)(dirname=%s)(type=directory)',
-                          account, dir);
-    } else {
-        filter += sprintf('(owner=%s)(dirname=%s)(type=object)',
-                          account, dir);
+    if ((hasDir || hasObj) && !(hasDir && hasObj)) {
+        filter.addFilter(new morayFilter.EqualityFilter({
+            attribute: 'type',
+            value: (hasDir ? 'directory' : 'object')
+        }));
     }
 
     var marker = req.params.marker;
     var reverse = req.params.sort_order === 'reverse';
     var tsort = req.params.sort === 'mtime';
 
-    if (marker) {
-        if (tsort) {
-            marker = new Date(marker).getTime();
-            if (!marker) {
-                ee = new EventEmitter();
-                setImmediate(function () {
-                    ee.emit('error',
-                            new InvalidParameterError('marker',
-                                                      req.params.marker));
-                });
-                return (ee);
-            }
-
-            if (reverse) {
-                filter += sprintf('(_mtime>=%s)', marker);
-            } else {
-                filter += sprintf('(_mtime<=%s)', marker);
-            }
-        } else {
-            if (reverse) {
-                filter += sprintf('(name<=%s)', marker);
-            } else {
-                filter += sprintf('(name>=%s)', marker);
-            }
-        }
-    }
-    filter += ')';
-
-
     var log = req.log;
     var opts = {
-        filter: filter,
         limit: l,
         requestId: req.getId(),
         sort: {},
@@ -677,9 +654,38 @@ function readdir(dir, req) {
         }
     }
 
+    if (marker) {
+        if (tsort) {
+            var mtime = Date.parse(marker);
+            if (Number.isFinite(mtime)) {
+                marker = mtime.toString();
+            } else {
+                ee = new EventEmitter();
+                setImmediate(function () {
+                    ee.emit('error',
+                            new InvalidParameterError('marker',
+                                                      req.params.marker));
+                });
+                return (ee);
+            }
+        }
+
+        var sortArgs = {
+            attribute: opts.sort.attribute,
+            value: marker
+        };
+        if (opts.sort.order === 'DESC') {
+            filter.addFilter(new morayFilter.GreaterThanEqualsFilter(sortArgs));
+        } else {
+            filter.addFilter(new morayFilter.LessThanEqualsFilter(sortArgs));
+        }
+    }
+
+    opts.filter = filter.toString();
+
     log.debug({
         dir: dir,
-        filter: filter
+        filter: opts.filter
     }, 'readdir: entered');
     var mreq = client.search(opts);
 
diff --git a/package.json b/package.json
index 4b3a25d..6933097 100644
--- a/package.json
+++ b/package.json
@@ -31,6 +31,7 @@
         "marlin": "git+ssh://git@github.com:joyent/manta-marlin.git#master",
         "mime": "1.2.11",
         "moray": "3.1.1",
+        "moray-filter": "1.0.0",
         "once": "1.3.0",
         "restify": "2.6.3",
         "vasync": "1.4.3",
-- 
2.21.0

